vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,464 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/csv_parser/parse_example_csv.h"
+
+#include "vw/core/best_constant.h"
+#include "vw/core/parse_args.h"
+#include "vw/core/parse_primitives.h"
+#include "vw/core/parser.h"
+
+#include <string>
+
+namespace VW
+{
+namespace parsers
+{
+namespace csv
+{
+int parse_examples(VW::workspace* all, io_buf& buf, VW::multi_ex& examples)
+{
+  bool keep_reading = all->custom_parser->next(*all, buf, examples);
+  return keep_reading ? 1 : 0;
+}
+
+void parser::handling_csv_separator(VW::workspace& all, std::string& str, const std::string& name)
+{
+  if (str.length() > 1)
+  {
+    char result = str[0];
+    if (str[0] == '\\')
+    {
+      switch (str[1])
+      {
+        // Allow to specify \t as tabs
+        // As pressing tabs usually means auto completion
+        case 't':
+          result = '\t';
+          break;
+        default:
+          break;
+      }
+    }
+
+    if ((result != str[0] && str.length() > 2) || result == str[0])
+    {
+      all.logger.err_warn(
+          "Multiple characters passed as {}, only the first one '{}' will be "
+          "read and the rest will be ignored.",
+          name, result);
+    }
+    str = result;
+  }
+}
+
+void parser::set_parse_args(VW::config::option_group_definition& in_options, parser_options* parsed_options)
```

Comment:
Would be preferable for this to be a reference since it should never be null
```suggestion
void parser::set_parse_args(VW::config::option_group_definition& in_options, parser_options& parsed_options)
```